vector/src/main/java/im/vector/riotx/features/discovery/change/SetIdentityServerViewModel.kt

```diff
@@ -0,0 +1,154 @@
+                    homeServerUrl = session.sessionParams.homeServerUrl,
+                    defaultIdentityServerUrl = session.identityService().getDefaultIdentityServer()
+            )
+        }
+
+        @JvmStatic
+        override fun create(viewModelContext: ViewModelContext, state: SetIdentityServerState): SetIdentityServerViewModel? {
+            val fragment: SetIdentityServerFragment = (viewModelContext as FragmentViewModelContext).fragment()
+            return fragment.viewModelFactory.create(state)
+        }
+
+        fun sanitatizeBaseURL(baseUrl: String): String {
+            var baseUrl1 = baseUrl
+            if (!baseUrl1.startsWith("http://") && !baseUrl1.startsWith("https://")) {
+                baseUrl1 = "https://$baseUrl1"
+            }
+            return baseUrl1
+        }
+    }
+
+    var currentWantedUrl: String? = null
+        private set
+
+    private val userLanguage = stringProvider.getString(R.string.resources_language)
+
+    override fun handle(action: SetIdentityServerAction) {
+        when (action) {
+            SetIdentityServerAction.UseDefaultIdentityServer   -> useDefault()
+            is SetIdentityServerAction.UseCustomIdentityServer -> usedCustomIdentityServerUrl(action)
+        }.exhaustive
+    }
+
+    private fun useDefault() = withState { state ->
+        state.defaultIdentityServerUrl?.let { doChangeIdentityServerUrl(it) }
+    }
+
+    private fun usedCustomIdentityServerUrl(action: SetIdentityServerAction.UseCustomIdentityServer) {
+        doChangeIdentityServerUrl(action.url)
+    }
+
+    private fun doChangeIdentityServerUrl(url: String) {
+        var baseUrl = url
+        if (baseUrl.isEmpty()) {
+            _viewEvents.post(SetIdentityServerViewEvents.Failure(R.string.settings_discovery_please_enter_server))
+            return
+        }
+        baseUrl = sanitatizeBaseURL(baseUrl)
+
+        currentWantedUrl = baseUrl
+
+        _viewEvents.post(SetIdentityServerViewEvents.Loading())
+
+        // First ping the identity server v2 API
+        mxSession.identityService().isValidIdentityServer(baseUrl, object : MatrixCallback<Unit> {
+            override fun onSuccess(data: Unit) {
+                // Ok, next step
+                checkTerms(baseUrl)
+            }
+
+            override fun onFailure(failure: Throwable) {
+                if (failure is IdentityServiceError.OutdatedIdentityServer) {
+                    _viewEvents.post(SetIdentityServerViewEvents.Failure(R.string.identity_server_error_outdated_identity_server))
+                } else {
+                    _viewEvents.post(SetIdentityServerViewEvents.Failure(R.string.settings_discovery_bad_identity_server))
+                    _viewEvents.post(SetIdentityServerViewEvents.OtherFailure(failure))
+                }
+            }
+        })
+    }
+
+    private fun checkTerms(baseUrl: String) {
+        mxSession.getTerms(TermsService.ServiceType.IdentityService,
```

Comment:
Could be cleaner with a viewModelScop.launch(Dispather..)
will avoid the call back {{{{

Reply:
Yes indeed. We will do that when we will remove the MatrixCallback class, right?

Reply:
Yes, but we should perform this change globally in the app, right?